src/Passes/Source/RuleTransformationPass/RulePass.cpp

```diff
@@ -0,0 +1,863 @@
+    {
+        active_pieces_.insert(input);
+        return input;
+    }
+
+    bool RuleTransformationPass::runOnFunction(llvm::Function& function, InstructionModifier const& modifier)
+    {
+        if (function.isDeclaration())
+        {
+            return false;
+        }
+
+        if (depth_ >= config_.maxRecursion())
+        {
+            llvm::outs() << "Exceed max recursion of " << config_.maxRecursion() << "\n";
+            return false;
+        }
+        ++depth_;
+
+        // Keep track of instructions scheduled for deletion
+        DeletableInstructions schedule_instruction_deletion;
+
+        // Block queue
+        std::deque<llvm::BasicBlock*>         queue;
+        std::unordered_set<llvm::BasicBlock*> blocks_queued;
+        queue.push_back(&function.getEntryBlock());
+        blocks_queued.insert(&function.getEntryBlock());
+
+        // Executing the modifier on the function itsel
+        modifier(&function, schedule_instruction_deletion);
+
+        while (!queue.empty())
+        {
+            auto& basic_block = *(queue.front());
+            queue.pop_front();
+
+            // Executing the modifier on the block
+            modifier(&basic_block, schedule_instruction_deletion);
+
+            for (auto& instr : basic_block)
+            {
+                // Modifying instruction as needed
+                auto instr_ptr = modifier(&instr, schedule_instruction_deletion);
+
+                // In case the instruction was scheduled for deletion
+                if (instr_ptr == nullptr)
+                {
+                    continue;
+                }
+
+                // Checking if we are calling a function
+                auto call_instr = llvm::dyn_cast<llvm::CallBase>(instr_ptr);
+                if (call_instr != nullptr)
+                {
+                    auto callee_function = call_instr->getCalledFunction();
+                    if (callee_function != nullptr && !callee_function->isDeclaration())
+                    {
+                        runOnFunction(*callee_function, modifier);
+                    }
+                }
+
+                // Following the branches to their basic blocks
+                auto* br_instr = llvm::dyn_cast<llvm::BranchInst>(&instr);
+                if (br_instr != nullptr)
+                {
+                    for (uint32_t i = 0; i < br_instr->getNumOperands(); ++i)
+                    {
+                        // TODO(tfr): This may not work on multi path branches (conditional)
+                        // as we may accidentality add the final path (contains qubit release)
+                        // and we cannot make assumptions since optimisation may have rearranged
+                        // everything. In this case, we should revert to the order they appear in the
+                        // function
+                        auto bb = llvm::dyn_cast<llvm::BasicBlock>(br_instr->getOperand(i));
```

Comment:
```suggestion
                        // TODO(tfr): This may not work on multi path branches (conditional)
                        // as we may accidentaly add the final path (contains qubit release)
                        // and we cannot make assumptions since optimisation may have rearranged
                        // everything. In this case, we should revert to the order they appear in the
                        // function
                        auto bb = llvm::dyn_cast<llvm::BasicBlock>(br_instr->getOperand(i));
```

Reply:
Fixed.